dpnp/backend/extensions/lapack/heevd_batch.cpp

```diff
@@ -0,0 +1,207 @@
+//*****************************************************************************
+// Copyright (c) 2024, Intel Corporation
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+// - Redistributions of source code must retain the above copyright notice,
+//   this list of conditions and the following disclaimer.
+// - Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//*****************************************************************************
+
+#include "heevd_batch.hpp"
+#include "common_helpers.hpp"
+
+namespace dpnp::extensions::lapack
+{
+namespace mkl_lapack = oneapi::mkl::lapack;
+namespace type_utils = dpctl::tensor::type_utils;
+
+template <typename T, typename RealT>
+static sycl::event heevd_batch_impl(sycl::queue &exec_q,
+                                    const oneapi::mkl::job jobz,
+                                    const oneapi::mkl::uplo upper_lower,
+                                    const std::int64_t batch_size,
+                                    const std::int64_t n,
+                                    char *in_a,
+                                    char *out_w,
+                                    const std::vector<sycl::event> &depends)
+{
+    type_utils::validate_type_for_device<T>(exec_q);
+    type_utils::validate_type_for_device<RealT>(exec_q);
+
+    T *a = reinterpret_cast<T *>(in_a);
+    RealT *w = reinterpret_cast<RealT *>(out_w);
+
```

Comment:
Do we need to have any special handling in case of `batch_size == 0`?
Is it possible that `batch_size == 0` passed into `heevd_batch_impl`?

Reply:
I think this check is unnecessary because in python we have `a.size == 0` check and `batch_size=0` can not be passed to `heevd_batch_impl or syevd_batch_impl`

Reply:
I have checked `heevd_batch_imp`l and `syevd_batch_impl` with `batch_size=0` and they work correctly so I do not think an additional check is needed for this case.

Reply:
We no need to allocate memory for scratchpad in case of `batch_size == 0`

Reply:
My check `helper::check_zeros_shape` in `evd_batch_func` ensures that `batch_size` and `n` are non-zero and returns empty events otherwise.
I added a comment with this and I think the additional checks in `syevd_batch_impl` and `heevd_batch_impl` do not make sense